### DISCUSSION TITLE
How to destroy the store instance in React?

### DISCUSSION THREAD

**Discussion by ChestnutYYY** [pmndrs/zustand#1746] [Q&A]
**Question 1:**
  I write a table store with some state, like  `page`,`pageSize` and so on ...
  Can `zustand store(table store)` be destroyed automatically when I switch routes?  or has some method in `zustand` that I didn't found.
  
  actually, when i switch routes,  These state are still the value of the last operation, not the initial value.

**Question 2:**
  In actual scenarios, the operation logic of all tables can abstract the same logic, so how can a table store have multiple instances.
  right now my solution is to iterate over a set of keys to generate different stores, but I don't think it's a good solution

I was wondering if anyone has a good solution.
Thank you so much

**Comment by dbritto-dev** [comment]
@rllxxx of course, let me answer your questions:

- 1. Yes, we shall reset the store using the `useEffect()` hook clean up function. Here's an example:

```ts
const location = useLocation()
const { resetStore } = useStore()

useEffect(() => {
  return () => {
    // Set the store to initial state
    resetStore()
  }
}, [location])
```

- 2. We shall create the store dynamically. Here's and example:

```ts
import { create } from 'zustand'

const createUsePaginationStore = (initialState = { page: 1, pageSize: 25 }) => {
  return create((set) => ({
    ...initialState,
    nextPage: () => set((state) => ({ page: state.page + 1 })),
    prevPage: () => set((state) => ({ page: state.page - 1 })),
    resetStore: () => set(initialState),
  }))
}


const usePaginationStore1 = createUsePaginationStore()
const usePaginationStore2 = createUsePaginationStore()
```

**Reply by ChestnutYYY** [reply]
First of all,  Thank you anwer!!!
About **Question2**， I wonder if there are other solutions. For example, can I use middleware  to solve the problem?

**Reply by dbritto-dev** [reply]
> First of all,  Thank you anwer!!!
> About **Question2**， I wonder if there are other solutions. For example, can I use middleware  to solve the problem?

The middlewares are for something different like adding logging or persisting data. Multiple instances is equal to multiple stores.

**Reply by dbritto-dev** [reply]
Let's say we want to also have a way to have something like an atom in that case we could do something like this:

```ts
const createGetUsePaginationStore = () => {
  const paginationStores = new Map()

  return (paginationStoreKey) => {
    if (!paginationStores.has(paginationStoreKey)) {
      paginationStores.set(paginationStoreKey, createUsePaginationStore())
    }
    
    return paginationStores.get(paginationStoreKey)
  }
}
```

```ts
const getUsePaginationStore = createGetUsePaginationStore()
const mainPaginationStore = getUsePaginationStore('mainPaginationStore')
```

**Comment by vaynevayne** [comment]
You want to use zustand instead of useState, which is only available to the current component, not sharing, right?

```jsx
// before 
const Demo =()=>{
const [state,setState] = useState()
return null
}

// after
const Demo =()=>{
const [state,setState] = useZustand() // fake hook
return null
}
```

**Reply by ChestnutYYY** [reply]
Hi guys, I'am confused by your answer,

About my question, I feel like [dbritto-dev](https://github.com/dbritto-dev) answer, like `const usePaginationStore1 = createUsePaginationStore()
const usePaginationStore2 = createUsePaginationStore()`  which is not easy to use, due to need to define usePaginationStore1, usePaginationStore2 and so on..

For me, I don't like this repetitive definition

**Reply by dbritto-dev** [reply]
@rllxxx here's you go https://github.com/pmndrs/zustand/discussions/1746#discussioncomment-5591424
